File: debug_config.py
import yaml

# Load the configuration
with open('config_global_20_agents.yaml', 'r') as file:
    config = yaml.safe_load(file)

print("Loaded configuration successfully")
print(f"Number of user agents: {len(config['agents']['user_agents'])}")

# Check each user agent
for i, agent in enumerate(config['agents']['user_agents']):
    print(f"\nAgent {i}:")
    if 'attributes' in agent:
        print(f"  Attributes: {agent['attributes']}")
        if 'is_miner' in agent['attributes']:
            print(f"  is_miner attribute value: '{agent['attributes']['is_miner']}'")
            print(f"  is_miner type: {type(agent['attributes']['is_miner'])}")
        else:
            print("  No is_miner attribute found")
    else:
        print("  No attributes found")

    # Check if it should be a miner
    is_miner = False
    if 'attributes' in agent and 'is_miner' in agent['attributes']:
        is_miner_value = agent['attributes']['is_miner']
        if isinstance(is_miner_value, str):
            is_miner = is_miner_value.lower() in ['true', '1', 'yes', 'on']
        elif isinstance(is_miner_value, bool):
            is_miner = is_miner_value

    print(f"  Should be miner: {is_miner}")

File: debug_config.rs
use std::fs::File;
use serde_yaml;
use std::collections::HashMap;

#[derive(Debug, serde::Deserialize)]
struct UserAgentConfig {
    daemon: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    wallet: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    user_script: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    attributes: Option<HashMap<String, String>>,
}

impl UserAgentConfig {
    /// Check if this agent is a miner based on attributes
    /// Returns true if "is_miner" is found in attributes and is "true" or true
    /// Returns false if "is_miner" is not found or is "false" or false
    pub fn is_miner_value(&self) -> bool {
        if let Some(attrs) = &self.attributes {
            if let Some(is_miner_value) = attrs.get("is_miner") {
                // Handle string representations
                match is_miner_value.to_lowercase().as_str() {
                    "true" | "1" | "yes" | "on" => return true,
                    "false" | "0" | "no" | "off" => return false,
                    _ => {} // Continue to check other formats
                }

                // Try to parse as boolean directly
                if let Ok(parsed_bool) = is_miner_value.parse::<bool>() {
                    return parsed_bool;
                }
            }
        }
        false
    }
}

#[derive(Debug, serde::Deserialize)]
struct AgentDefinitions {
    #[serde(skip_serializing_if = "Option::is_none")]
    user_agents: Option<Vec<UserAgentConfig>>,
}

#[derive(Debug, serde::Deserialize)]
struct Config {
    general: serde_yaml::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    network: Option<serde_yaml::Value>,
    agents: AgentDefinitions,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Load the configuration
    let file = File::open("config_global_20_agents.yaml")?;
    let config: Config = serde_yaml::from_reader(file)?;

    println!("Loaded configuration successfully");
    println!("Number of user agents: {}", config.agents.user_agents.as_ref().unwrap().len());

    // Check each user agent
    if let Some(user_agents) = &config.agents.user_agents {
        for (i, agent) in user_agents.iter().enumerate() {
            println!("Agent {}: is_miner = {}", i, agent.is_miner_value());
            if let Some(attrs) = &agent.attributes {
                println!("  Attributes: {:?}", attrs);
                if let Some(is_miner_val) = attrs.get("is_miner") {
                    println!("  is_miner attribute value: '{}'", is_miner_val);
                } else {
                    println!("  No is_miner attribute found");
                }
            } else {
                println!("  No attributes found");
            }
        }
    }

    Ok(())
}